src/main/java/stirling/software/SPDF/service/CustomPDDocumentFactory.java

```diff
@@ -4,142 +4,345 @@
+
+        // If free memory is critically low, always use file-based caching
+        if (freeMemoryPercent < MIN_FREE_MEMORY_PERCENTAGE
+                || actualFreeMemory < MIN_FREE_MEMORY_BYTES) {
+            log.info(
+                    "Low memory detected ({}%), forcing file-based cache",
+                    String.format("%.2f", freeMemoryPercent));
+            cacheFunction = createScratchFileCacheFunction(MemoryUsageSetting.setupTempFileOnly());
+        } else if (contentSize < SMALL_FILE_THRESHOLD) {
+            log.info("Using memory-only cache for small document ({}KB)", contentSize / 1024);
+            cacheFunction = IOUtils.createMemoryOnlyStreamCache();
+        } else {
+            log.info("Using file-based cache");
+            cacheFunction = createScratchFileCacheFunction(MemoryUsageSetting.setupTempFileOnly());
+        }
+
+        PDDocument document;
+        if (source instanceof File file) {
+            document = loadFromFile(file, contentSize, cacheFunction);
+        } else if (source instanceof byte[] bytes) {
+            document = loadFromBytes(bytes, contentSize, cacheFunction);
+        } else {
+            throw new IllegalArgumentException("Unsupported source type: " + source.getClass());
+        }
+
+        postProcessDocument(document);
         return document;
     }
 
+    private StreamCacheCreateFunction createScratchFileCacheFunction(MemoryUsageSetting settings) {
+        return () -> {
+            try {
+                return new ScratchFile(settings);
+            } catch (IOException e) {
+                throw new RuntimeException("ScratchFile initialization failed", e);
+            }
+        };
+    }
+
+    private void postProcessDocument(PDDocument doc) throws IOException {
+        pdfMetadataService.setDefaultMetadata(doc);
+        removePassword(doc);
+    }
+
+    private PDDocument loadFromFile(File file, long size, StreamCacheCreateFunction cache)
+            throws IOException {
+        if (size >= EXTREMELY_LARGE_THRESHOLD) {
+            log.info("Loading extremely large file via buffered access");
+            return Loader.loadPDF(new RandomAccessReadBufferedFile(file), "", null, null, cache);
+        }
+        return Loader.loadPDF(file, "", null, null, cache);
+    }
+
+    private PDDocument loadFromBytes(byte[] bytes, long size, StreamCacheCreateFunction cache)
+            throws IOException {
+        if (size >= SMALL_FILE_THRESHOLD) {
+            log.info("Writing large byte array to temp file");
+            Path tempFile = createTempFile("pdf-bytes-");
+            try {
+                Files.write(tempFile, bytes);
+                return Loader.loadPDF(tempFile.toFile(), "", null, null, cache);
+            } finally {
+                cleanupFile(tempFile);
+            }
+        }
+        return Loader.loadPDF(bytes, "", null, null, cache);
+    }
+
+    public PDDocument createNewDocument(MemoryUsageSetting settings) throws IOException {
```

Comment:
I'd make this `private` since the other overloaded method is the only one that uses it

Reply:
plan to keep this open, i think we will be using this for some features going forward

Reply:
Will leave for now, i think this is only half done, need to propagate this loading to other classes and this may be handy